File: weapon.py
"""Data-driven weapons (stats from data/weapons.json) + WeaponManager."""
from __future__ import annotations

import math
import random

import pygame

import settings as S
from utils import load_json

WEAPON_DATA: dict[str, dict] = load_json(S.WEAPONS_FILE, {
    "pistol": {"name": "PISTOL", "damage": 25, "magazine": 12, "fire_rate": 0.3,
               "reload_time": 1.2, "bullet_speed": 1000, "spread_deg": 2.5,
               "pellets": 1, "critical_chance": 0.10, "critical_multiplier": 2.0,
               "auto": False, "price": 0, "start_reserve": 96},
})
WEAPON_ORDER = ["pistol", "shotgun", "smg", "rifle", "sniper",
                 "flamethrower", "plasma", "crossbow"]


class Weapon:
    """Runtime weapon instance: holds ammo state for one weapon type."""

    def __init__(self, weapon_id: str) -> None:
        self.id = weapon_id
        data = WEAPON_DATA.get(weapon_id)
        if data is None:                       # fallback: clone pistol
            data = dict(WEAPON_DATA["pistol"])
            self.id = "pistol"
        self.name: str = data.get("name", weapon_id.upper())
        self.damage: float = float(data.get("damage", 25))
        self.magazine_size: int = int(data.get("magazine", 12))
        self.fire_rate: float = float(data.get("fire_rate", 0.3))
        self.reload_time: float = float(data.get("reload_time", 1.2))
        self.bullet_speed: float = float(data.get("bullet_speed", 1000))
        self.spread_deg: float = float(data.get("spread_deg", 3))
        self.pellets: int = int(data.get("pellets", 1))
        self.critical_chance: float = float(data.get("critical_chance", 0.1))
        self.critical_multiplier: float = float(data.get("critical_multiplier", 2))
        self.auto: bool = bool(data.get("auto", False))
        self.price: int = int(data.get("price", 0))

        self.ammo = self.magazine_size
        self.reserve: int = int(data.get("start_reserve", 48))
        self.cooldown = 0.0
        self.reloading = False
        self.reload_timer = 0.0
        self.reload_total = self.reload_time

    # -------------------------------------------------------------- combat -
    def can_fire(self, want_held_down: bool) -> bool:
        if not want_held_down and not self.auto:
            return False
        return (self.cooldown <= 0 and not self.reloading
                and self.ammo > 0)

    def fire(self, base_angle: float, damage_mult: float = 1.0,
             crit_bonus: float = 0.0, crit_mult_bonus: float = 0.0
             ) -> list[dict]:
        """Fire one shot; returns bullet specs (angle, damage, crit)."""
        self.ammo -= 1
        self.cooldown = self.fire_rate
        shots: list[dict] = []
        for _ in range(max(1, self.pellets)):
            angle = base_angle + math.radians(
                random.uniform(-self.spread_deg, self.spread_deg))
            crit = random.random() < min(0.9, self.critical_chance + crit_bonus)
            mult = self.critical_multiplier + crit_mult_bonus if crit else 1.0
            shots.append({
                "angle": angle,
                "damage": self.damage * damage_mult * mult,
                "crit": crit,
                "speed": self.bullet_speed,
            })
        return shots

    def start_reload(self, mult: float = 1.0) -> bool:
        if self.reloading or self.ammo >= self.magazine_size or self.reserve <= 0:
            return False
        self.reloading = True
        self.reload_total = self.reload_time * mult
        self.reload_timer = self.reload_total
        return True

    def update(self, dt: float) -> None:
        self.cooldown = max(0.0, self.cooldown - dt)
        if self.reloading:
            self.reload_timer -= dt
            if self.reload_timer <= 0:
                need = self.magazine_size - self.ammo
                take = min(need, self.reserve)
                self.ammo += take
                self.reserve -= take
                self.reloading = False

    def add_reserve(self, amount: int) -> None:
        self.reserve += max(0, int(amount))


class WeaponManager:
    """Owns the player's unlocked weapons and the active selection."""

    def __init__(self, unlocked_ids: list[str] | None = None) -> None:
        self.weapons: dict[str, Weapon] = {}
        for wid in unlocked_ids or ["pistol"]:
            self.give(wid)
        if not self.weapons:
            self.give("pistol")
        self.current_id = next(iter(self.weapons))
        # Optional callback: host game sets this to receive switch events
        # (slot_index, weapon_id) for HUD feedback. Keeping it on the
        # manager avoids a back-reference cycle into Player.
        self.switch_notify = None

    def give(self, weapon_id: str) -> bool:
        """Add a new weapon type. Returns True if it was newly added."""
        if weapon_id in self.weapons:
            return False
        self.weapons[weapon_id] = Weapon(weapon_id)
        return True

    @property
    def current(self) -> Weapon:
        return self.weapons[self.current_id]

    def unlocked_in_order(self) -> list[str]:
        """Weapon ids unlocked by the player, in canonical display order."""
        return [w for w in WEAPON_ORDER if w in self.weapons]

    def select_slot(self, slot: int) -> bool:
        """Legacy alias kept for backwards compatibility."""
        return self.switch_weapon(slot, announce=False)

    def switch_weapon(self, slot: int, announce: bool = True) -> bool:
        """Switch active weapon by 1..5 slot index.

        Validates the slot, falls back gracefully when the slot is locked,
        cancels any in-progress reload so the new weapon starts clean, and
        notifies the host game (via ``switch_notify``) so it can show a
        short feedback banner. Returns True iff the active weapon changed.
        """
        if not isinstance(slot, int):
            return False
        ids = self.unlocked_in_order()
        if not (1 <= slot <= len(ids)):
            return False  # slot locked / out of range → keep current weapon
        new_id = ids[slot - 1]
        old_id = self.current_id
        if new_id == old_id:
            # Same weapon pressed again: still cancel a stale reload so
            # the magazine state stays consistent.
            cur = self.current
            if cur.reloading:
                cur.reloading = False
                cur.reload_timer = 0.0
            return False
        # Cancel reload on the OLD weapon so we don't carry its timer over
        # into the new one and so ammo doesn't get applied to the wrong gun.
        old = self.weapons.get(old_id)
        if old is not None and old.reloading:
            old.reloading = False
            old.reload_timer = 0.0
        # Defensive: also clear any stale reload on weapons the player
        # is *not* currently holding — prevents the previous weapon
        # from sneaking its reload-completion into the new weapon's ammo.
        for wid, wp in self.weapons.items():
            if wid != new_id and wp.reloading:
                wp.reloading = False
                wp.reload_timer = 0.0
        self.current_id = new_id
        if announce and self.switch_notify is not None:
            try:
                self.switch_notify(new_id, slot)
            except Exception:
                pass
        return True

    def cycle(self, direction: int = 1) -> None:
        ids = self.unlocked_in_order()
        if not ids:
            return
        idx = ids.index(self.current_id)
        self.switch_weapon(((idx + direction) % len(ids)) + 1)

    def update(self, dt: float) -> None:
        self.current.update(dt)

    def draw_muzzle_hint(self) -> None:  # kept for API symmetry / future use
        pass

File: web/src/game/weapon.ts
// src/game/weapon.ts
// Data-driven weapons + WeaponManager. Mirrors weapon.py.

import { BULLET_LIFETIME } from "./settings";
import type { WeaponData } from "./data";
import { modMultiplier } from "./mods";

export const WEAPON_ORDER: readonly string[] = [
  "pistol",
  "shotgun",
  "smg",
  "rifle",
  "sniper",
  "flamethrower",
  "plasma",
  "crossbow",
];

export interface FireResult {
  angle: number;
  damage: number;
  crit: boolean;
  speed: number;
  /** Seconds this shot may fly = range / speed. */
  lifetime: number;
  /** Elemental projectile behaviour for this shot. */
  elem?: "fire" | "plasma" | "pierce";
  /** Projectile radius for this shot. */
  radius: number;
}

export class Weapon {
  id: string;
  name: string;
  damage: number;
  /** Effective (mod-adjusted) stats used by the rest of the game. */
  magazineSize: number;
  fireRate: number;
  reloadTime: number;
  bulletSpeed: number;
  /** Max reach in px before the projectile expires. */
  range: number;
  elem?: "fire" | "plasma" | "pierce";
  radius: number;
  spreadDeg: number;
  pellets: number;
  criticalChance: number;
  criticalMultiplier: number;
  auto: boolean;
  price: number;

  /** Un-modded base stats, kept so mods can be recomputed idempotently. */
  private baseMagazineSize: number;
  private baseReloadTime: number;
  private baseRange: number;
  private baseSpreadDeg: number;
  /** Mod ids attached to this weapon (persisted via SaveData.weapon_upgrades). */
  mods: string[];

  ammo: number;
  reserve: number;
  cooldown = 0;
  reloading = false;
  reloadTimer = 0;
  reloadTotal = 0;

  constructor(id: string, data: Record<string, WeaponData>, mods: string[] = []) {
    const d = data[id] ?? data["pistol"]!;
    this.id = data[id] ? id : "pistol";
    this.name = d.name;
    this.damage = d.damage;
    this.baseMagazineSize = d.magazine;
    this.fireRate = d.fire_rate;
    this.baseReloadTime = d.reload_time;
    this.bulletSpeed = d.bullet_speed;
    this.baseRange = d.range ?? this.bulletSpeed * BULLET_LIFETIME; // legacy fallback
    this.elem = d.elem;
    this.radius = d.bullet_radius ?? 4;
    this.baseSpreadDeg = d.spread_deg;
    this.pellets = d.pellets;
    this.criticalChance = d.critical_chance;
    this.criticalMultiplier = d.critical_multiplier;
    this.auto = d.auto;
    this.price = d.price;

    this.mods = [...mods];
    this.magazineSize = this.baseMagazineSize;
    this.reloadTime = this.baseReloadTime;
    this.range = this.baseRange;
    this.spreadDeg = this.baseSpreadDeg;
    this.recomputeMods();

    this.ammo = this.magazineSize;
    this.reserve = d.start_reserve;
  }

  /** Re-derive the effective stats from base values + attached mods. */
  recomputeMods(): void {
    this.magazineSize = Math.round(this.baseMagazineSize * modMultiplier(this.mods, "magazine"));
    this.reloadTime = this.baseReloadTime * modMultiplier(this.mods, "reload");
    this.range = this.baseRange * modMultiplier(this.mods, "range");
    this.spreadDeg = this.baseSpreadDeg * modMultiplier(this.mods, "spread");
  }

  /** Attach `modId`; returns false if already equipped. */
  addMod(modId: string): boolean {
    if (this.mods.includes(modId)) return false;
    this.mods.push(modId);
    this.recomputeMods();
    return true;
  }

  canFire(wantHeldDown: boolean): boolean {
    if (!wantHeldDown && !this.auto) return false;
    return this.cooldown <= 0 && !this.reloading && this.ammo > 0;
  }

  fire(
    baseAngle: number,
    damageMult = 1,
    critBonus = 0,
    critMultBonus = 0,
  ): FireResult[] {
    this.ammo -= 1;
    this.cooldown = this.fireRate;
    const shots: FireResult[] = [];
    for (let i = 0; i < Math.max(1, this.pellets); i++) {
      const angle =
        baseAngle + (Math.random() * 2 - 1) * (this.spreadDeg * Math.PI) / 180;
      const crit = Math.random() < Math.min(0.9, this.criticalChance + critBonus);
      const mult = crit ? this.criticalMultiplier + critMultBonus : 1;
      shots.push({
        angle,
        damage: this.damage * damageMult * mult,
        crit,
        speed: this.bulletSpeed,
        lifetime: this.range / Math.max(1, this.bulletSpeed),
        elem: this.elem,
        radius: this.radius,
      });
    }
    return shots;
  }

  startReload(mult = 1): boolean {
    if (this.reloading) return false;
    if (this.ammo >= this.magazineSize) return false;
    if (this.reserve <= 0) return false;
    this.reloading = true;
    this.reloadTotal = this.reloadTime * mult;
    this.reloadTimer = this.reloadTotal;
    return true;
  }

  update(dt: number): void {
    this.cooldown = Math.max(0, this.cooldown - dt);
    if (this.reloading) {
      this.reloadTimer -= dt;
      if (this.reloadTimer <= 0) {
        const need = this.magazineSize - this.ammo;
        const take = Math.min(need, this.reserve);
        this.ammo += take;
        this.reserve -= take;
        this.reloading = false;
      }
    }
  }

  addReserve(amount: number): void {
    this.reserve += Math.max(0, Math.floor(amount));
  }
}

export class WeaponManager {
  weapons: Record<string, Weapon> = {};
  currentId: string = "pistol";
  data: Record<string, WeaponData>;
  weaponMods: Record<string, string[]>;

  constructor(
    unlocked: string[] = ["pistol"],
    data: Record<string, WeaponData>,
    weaponMods: Record<string, string[]> = {},
  ) {
    this.data = data;
    this.weaponMods = weaponMods;
    for (const id of unlocked) this.give(id);
    if (Object.keys(this.weapons).length === 0) this.give("pistol");
    this.currentId = Object.keys(this.weapons)[0]!;
  }

  give(id: string): boolean {
    if (this.weapons[id]) return false;
    this.weapons[id] = new Weapon(id, this.data, this.weaponMods[id]);
    return true;
  }

  /** Attach `modId` to an already-owned weapon. Returns false if not owned or already equipped. */
  applyMod(id: string, modId: string): boolean {
    const w = this.weapons[id];
    if (!w) return false;
    return w.addMod(modId);
  }

  get current(): Weapon {
    const w = this.weapons[this.currentId];
    if (w) return w;
    // Defensive: currentId points at a weapon that isn't owned (can happen
    // after restoring a save whose unlocked list drifted from currentId).
    // Fall back to the first owned weapon in canonical order and re-anchor.
    const fallback = WEAPON_ORDER.find((wid) => this.weapons[wid]);
    if (fallback) {
      this.currentId = fallback;
      return this.weapons[fallback]!;
    }
    // Last resort: hand back the first map entry (constructor guarantees
    // at least "pistol" exists, so this can only be reached if the map
    // itself is empty, which never happens).
    return Object.values(this.weapons)[0]!;
  }

  selectSlot(slot: number): boolean {
    const ids = WEAPON_ORDER.filter((w) => this.weapons[w]);
    if (slot >= 1 && slot <= ids.length) {
      this.currentId = ids[slot - 1]!;
      return true;
    }
    return false;
  }

  cycle(direction = 1): void {
    const ids = WEAPON_ORDER.filter((w) => this.weapons[w]);
    if (ids.length === 0) return;
    const idx = ids.indexOf(this.currentId);
    // If currentId is not in the owned list (e.g. corrupted save), start
    // from the first slot instead of producing -1.
    const safeIdx = idx === -1 ? 0 : idx;
    this.currentId = ids[(safeIdx + direction + ids.length) % ids.length]!;
  }

  update(dt: number): void {
    this.current.update(dt);
  }
}
